Media Stats should show the media element's source

jernoble opened the pull request:
#### 0ee37c4333af683443d1934d4cbb137bfcc7cb26
<pre>
Media Stats should show the media element&apos;s source
<a href="https://bugs.webkit.org/show_bug.cgi?id=256657">https://bugs.webkit.org/show_bug.cgi?id=256657</a>
rdar://109219273

Reviewed by Eric Carlson.

Media Stats currently shows a lot of great information about the codecs and properties of the
media element&apos;s video and audio tracks. Also useful would be information about the element itself,
such as whether it&apos;s current source was a file, stream, media source, etc.

Add a new attribute to MediaControlsHost which vends this information up to the media controls. Add
support for detecting whether the AVURLAsset is backed by a HLS stream to MediaplayerPrivateAVFoundationObjC.

* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
(WebCore::MediaControlsHost::sourceType const):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.idl:
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.setShowingStats):
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(WK_ucfirst):
* Source/WebCore/en.lproj/modern-media-controls-localized-strings.js:
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::movieLoadType const):
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
(WebCore::MediaPlayerPrivateAVFoundation::isHLS const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::assetStatus const):
(WebCore::assetMetadataKeyNames):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/264293@main">https://commits.webkit.org/264293@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34667271eaae7098e492f8ec036e6c4281373651

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6317 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6693 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7892 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6639 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6318 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6734 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6463 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9544 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6429 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6435 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5721 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7963 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3907 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13586 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5771 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5784 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8035 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6266 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5123 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5676 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1719 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9831 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6047 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
```diff
@@ -105,6 +105,18 @@ class MediaControlsHost final : public RefCounted<MediaControlsHost>, public Can
 #endif // ENABLE(MEDIA_CONTROLS_CONTEXT_MENUS)
 #endif // ENABLE(MODERN_MEDIA_CONTROLS)
 
+    enum SourceType {
```
```suggestion
    enum class SourceType : uint8_t {
```

jernoble replied:
Yes.

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -4137,6 +4137,24 @@ void HTMLMediaElement::pauseInternal()
+
+bool HTMLMediaElement::hasManagedMediaSource() const
+{
+#if ENABLE(MANAGED_MEDIA_SOURCE)
```
Should probably be:
```
#if ENABLE(MANAGED_MEDIA_SOURCE)
    return is<ManagedMediaSource>(m_mediaSource);
#else
    return false;
#endif
```

or

```
#if ENABLE(MANAGED_MEDIA_SOURCE)
    if (m_mediaSource)
        return is<ManagedMediaSource>(*m_mediaSource);
#endif
    return false;
```

cdumez replied:
`is<>` already does a null check for you when you give it a pointer instead of a reference.

jernoble committed:
Media Stats should show the media element's source
https://bugs.webkit.org/show_bug.cgi?id=256657
rdar://109219273

Reviewed by Eric Carlson.

Media Stats currently shows a lot of great information about the codecs and properties of the
media element's video and audio tracks. Also useful would be information about the element itself,
such as whether it's current source was a file, stream, media source, etc.

Add a new attribute to MediaControlsHost which vends this information up to the media controls. Add
support for detecting whether the AVURLAsset is backed by a HLS stream to MediaplayerPrivateAVFoundationObjC.

* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
(WebCore::MediaControlsHost::sourceType const):
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.idl:
* Source/WebCore/Modules/modern-media-controls/media/media-controller.js:
(MediaController.prototype.setShowingStats):
* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(WK_ucfirst):
* Source/WebCore/en.lproj/modern-media-controls-localized-strings.js:
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::movieLoadType const):
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
(WebCore::MediaPlayerPrivateAVFoundation::isHLS const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::assetStatus const):
(WebCore::assetMetadataKeyNames):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/264293@main
